### PR TITLE
Add GCP Secret Manager Backend

### LINKED ISSUE #5: Add support for GCP Secret Manager

### PR DESCRIPTION
**Pull request by DirtyCajunRice**
Fixes #5

### CODE REVIEW

File: pkg/apis/secretmanager/v1alpha1/externalsecret_types.go [hunk 1]
```diff
@@ -65,7 +65,9 @@ type ObjectReference struct {
 type KeyReference struct {
 	// The key in the generated secret to place fetched secret value into.
 	SecretKey string `json:"secretKey"`
-
+	// Namespace of the secret. Ignored if SecretStore
+	// +optional
+	Namespace *string `json:"namespace,omitempty"`
```
Comment (mcavoyk) [inline pkg/apis/secretmanager/v1alpha1/externalsecret_types.go:70]: ExternalSecrets are namespaced scope, they should only be able to create a secret in the same namespace as the ExternalSecret.

File: pkg/internal/gcp/gcp.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"google.golang.org/api/secretmanager/v1"
```
Comment (mcavoyk) [inline pkg/internal/gcp/gcp.go:23]: Isn't https://github.com/googleapis/google-cloud-go the more official SDK. https://github.com/googleapis/google-api-go-client is auto-gen'd.

File: pkg/internal/gcp/gcp.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	spec := g.store.GetSpec().GCP
+	if spec.AuthSecretRef == nil {
+		g.log.V(1).Info("no authentication defined. using environment variables")
+		g.client, err = secretmanager.NewService(ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	if spec.AuthSecretRef.JSON != nil && spec.AuthSecretRef.File != nil {
```
Comment (mcavoyk) [inline pkg/internal/gcp/gcp.go:113]: This logic should be in a validating webhook: ref https://github.com/itscontained/secret-manager/issues/6.

Should either remove or put a TODO to remove when adding validating webhook.

File: pkg/internal/store/base/store.go [hunk 4]
```diff
@@ -30,21 +34,23 @@ var _ store.Factory = &Default{}
...
-			return nil, fmt.Errorf("unable to setup Vault client: %w", err)
-		}
-		return vaultClient, nil
-	} else if storeSpec.AWS != nil {
-		awsClient, err := aws.New(ctx, kubeClient, store)
-		if err != nil {
-			return nil, fmt.Errorf("unable to setup AWS client: %w", err)
-		}
-		return awsClient, nil
+func (f *Default) New(ctx context.Context, genericStore smv1alpha1.GenericStore, kubeClient client.Client, _ client.Reader, namespace string, log logr.Logger) (store.Client, error) {
```
Comment (mcavoyk) [inline pkg/internal/store/base/store.go:37]: I would re-order so its - `(ctx, logger, store, kubeclient, namespace)`

File: pkg/apis/secretmanager/v1alpha1/secretstore_gcp_types.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	// The JSON secret key selector is used for authentication. If not set we fall-back to using
+	// `GOOGLE_APPLICATION_CREDENTIALS` or the default service account of the compute engine
+	// see: https://cloud.google.com/docs/authentication/production
+	// +optional
+	JSON *smmeta.SecretKeySelector `json:"json,omitempty"`
+	// The File string is used for authentication using a gcp credentials json file. If not set we fall-back to using
+	// `GOOGLE_APPLICATION_CREDENTIALS` or the default service account of the compute engine
+	// see: https://cloud.google.com/docs/authentication/production
+	// +optional
+	File *string `json:"file,omitempty"`
```
Comment (mcavoyk) [inline pkg/apis/secretmanager/v1alpha1/secretstore_gcp_types.go:43]: Thoughts on `filePath` to be more specific here. 

I'm also not sure this setting is needed as this already requires configuration of the controller to have the credential file mounted, feels like we are better off just letting them use default file location or ENV configuration than allowing overriding file location.

Comment (mcavoyk) [reply pkg/apis/secretmanager/v1alpha1/secretstore_gcp_types.go:43]: I guess to be more specific, configuring this requires also configuring the controller to already have the file at this location, and tightly couples controller and store. At that point it seems more natural to just use cloud provider configuration than overriding it in the store.

File: pkg/internal/gcp/gcp.go (hunk 2 above)
Comment (DirtyCajunRice) [reply pkg/internal/gcp/gcp.go:23]: * [docs](https://pkg.go.dev/cloud.google.com/go/secretmanager/apiv1#pkg-overview) show `cloud.google.com/go/secretmanager/apiv1` is autogenerated
* [docs](https://pkg.go.dev/google.golang.org/api/secretmanager/v1#pkg-overview) show `google.golang.org/api/secretmanager/v1` is not

File: pkg/internal/gcp/gcp.go (hunk 3 above)
Comment (DirtyCajunRice) [reply pkg/internal/gcp/gcp.go:113]: added

File: pkg/internal/store/base/store.go (hunk 4 above)
Comment (DirtyCajunRice) [reply pkg/internal/store/base/store.go]: re-ordered

File: pkg/apis/secretmanager/v1alpha1/externalsecret_types.go (hunk 1 above)
Comment (DirtyCajunRice) [reply pkg/apis/secretmanager/v1alpha1/externalsecret_types.go:70]: removed

File: pkg/apis/secretmanager/v1alpha1/secretstore_gcp_types.go (hunk 5 above)
Comment (DirtyCajunRice) [reply pkg/apis/secretmanager/v1alpha1/secretstore_gcp_types.go]: Renamed.

Reference: Decided in discord that existence is plausible in the event they want unique credentials for multiple stores mounted.

**Review by mcavoyk — APPROVED** [review]

**Pull request merged**